File: awslambda/iversity/handler.py
import json
import time
from os import sys, path
from datetime import datetime

sys.path.append(path.dirname(path.dirname(path.abspath(__file__))))
import utils


def lambda_handler(event, context):
    PROVIDER = 'Iversity'

    URL = "https://iversity.org/api/v1/courses"
    POST_URL = "https://search-coursehub-mmrw23yio2a4ylx2cgmx34eiyy.us-east-2.es.amazonaws.com/courses/course"
    SEARCH_URL = "https://search-coursehub-mmrw23yio2a4ylx2cgmx34eiyy.us-east-2.es.amazonaws.com/courses/_search"

    courses = utils.fetch_data(URL, 5, 30)
    if courses != False:
        count = 1
        add_count = 0
        start = time.time()
        total = len(courses)

        for course in courses:
            try:
                if 'Eng' in course['language']:
                    key = PROVIDER + '-'
                    key += str(course['id']) if 'id' in course else course['url']

                    if utils.search_data_elastic_search(SEARCH_URL, {"CourseId": key}):
                        print('[$] Already Processed', count, '/', total)
                        count += 1
                        continue

                    instructors = [{"InstructorId": ins['id'] if 'id' in ins else None,
                                    "InstructorName": ins['name'] if 'name' in ins else None,
                                    "ProfilePic": ins['image'] if 'image' in ins else None}
                                   for ins in course['instructors']] if 'instructors' in course else None

                    data = ({"CourseId": key, "CourseProvider": PROVIDER,
                             "Title": course["title"] if 'title' in course else None,
                             "Category": course["discipline"] if 'discipline' in course else None,
                             "CourseDuration": int(course["duration"]) if 'duration' in course and len(course["duration"]) != 0 else None,
                             "Paid": True if 'verifications' in course and course["verifications"][0]['price'] is not None else False,
                             "Price": int(course["verifications"][0]['price'].split(" ")[0]) if 'verifications' in course and course["verifications"][0]["price"] is not None else None,
                             "PriceCurrency": "Euro", "Instructors": instructors,
                             "URL": course["url"] if 'url' in course else None,
                             "Rating": 0, "Description": course["description"] if 'description' in course else None,
                             "CourseImage": course["image"] if "image" in course else None,
                             "StartDate": datetime.strptime(course["start_date"].split('T')[0], '%Y-%m-%d').isoformat().split('T')[0]
                             if "start_date" in course and course["start_date"] is not None else None,
                             "EndDate":datetime.strptime(course["end_date"].split('T')[0], '%Y-%m-%d').isoformat().split('T')[0]
                             if "end_date" in course and course["end_date"] is not None else None,
                             "SelfPaced": None})

                    utils.add_data_elastic_search(POST_URL, data)
                    add_count += 1
                else:
                    print('[X] Course text not in english', count, '/', total)
                count += 1
            except Exception as e:
                print('[%] Error occured ', count, '/', total, ':', e)


        print("\nCourse Catalog Update Complete")
        print("Total Courses Fetch:", total)
        print("Total Courses Added:", add_count)
        print('Total time taken:', round(time.time() - start, 2), 'sec\n')
        return True

if __name__ == '__main__':
    lambda_handler('', '')

File: awslambda/edx/handler.js
'use strict';

var dotenv = require('dotenv');
const fetch = require("node-fetch");
const { URLSearchParams } = require('url');
const helpers = require('./helpers');
dotenv.load();

module.exports.edxDaily = async (event, context) => {
  const params = new URLSearchParams();
  params.append('grant_type', 'client_credentials');
  params.append('client_id', process.env.EDX_CLIENT_ID);
  params.append('client_secret', process.env.EDX_CLIENT_SECRET);
  params.append('token_type', 'jwt');

  var accessToken;

  return new Promise((resolve, reject) => {
    fetch(process.env.EDX_GET_TOKEN, { method: 'POST', body: params }).then(response => {
      // console.log("Status: ", response.status);
      if (response.status === 200) {
        return response.json();
      } else {
        throw Error(response.status + ": " + response.statusText);
      }
    }).then(tokenData => {
      // console.log("Access Token: ", tokenData.access_token);
      accessToken = tokenData.access_token;
      return fetch(process.env.EDX_FETCH_COURSES, { headers: { 'Authorization': 'JWT ' + tokenData.access_token } });
    }).then(catalog => {
      // console.log("Response Status: ", catalog.status, " :: Response Status Text: ", catalog);
      if (catalog.status === 200) {
        return catalog.json();
      } else {
        throw Error(catalog.status + ": " + catalog.statusText);
      }
    }).then(catalogData => {
      var courses = catalogData.results.
        filter(course => course.key !== null && course.key !== '' && typeof course.course_runs !== 'undefined').
        map(course => ({
          "CourseId": "EDX-" + course.key, //id
          "CourseProvider": "EDX", //provider
          "Title": course.title, //title
          "Category": typeof course.subjects === 'undefined' ? [] : course.subjects.filter(subject => subject.slug !== 'undefined' && subject.slug !== null).map(subject => subject.slug), //category
          "CourseDuration": { "Value": course.course_runs[0].max_effort * course.course_runs[0].weeks_to_complete, "Unit": "hrs" }, //duration
          "Paid": typeof course.course_runs[0].seats === 'undefined' ? false : course.course_runs[0].seats.filter(seat => typeof seat.price !== 'undefined' && seat.price === "0.00").length === 0, //paid?
          "Price": typeof course.course_runs[0].seats === 'undefined' ? false : course.course_runs[0].seats.filter(seat => typeof seat.price !== 'undefined' && seat.price === "0.00").length !== 0 ? 0.0 : parseFloat(course.course_runs[0].seats.filter(seat => seat.price !== "0.00")[0].price), //price
          "PriceCurrency": typeof course.course_runs[0].seats === 'undefined' ? null : course.course_runs[0].seats[0].currency,
          "Instructors": (typeof course.course_runs[0].staff === 'undefined' || course.course_runs[0].staff.length === 0) ? [] : course.course_runs[0].staff.map(staff => ({ InstructorId: staff.uuid, InstructorName: staff.given_name + " " + staff.family_name, ProfilePic: staff.profile_image_url })),
          "URL": typeof course.course_runs[0].marketing_url === 'undefined' ? null : course.course_runs[0].marketing_url,
          "Rating": 0.0,
          "Description": typeof course.course_runs[0].full_description === 'undefined' ? null : course.course_runs[0].full_description,
          "CourseImage": (typeof course.course_runs[0].image === 'undefined' || typeof course.course_runs[0].image.src === 'undefined') ? null : course.course_runs[0].image.src,
          "StartDate": (typeof course.course_runs[0].start === 'undefined' || course.course_runs[0].start === null) ? null : (isNaN(Date.parse(course.course_runs[0].start)) ? null : new Date(Date.parse(course.course_runs[0].start)).getFullYear() + "-" + (parseInt(new Date(Date.parse(course.course_runs[0].start)).getMonth()) + 1) + "-" + new Date(Date.parse(course.course_runs[0].start)).getDate()),
          "EndDate": (typeof course.course_runs[0].end === 'undefined' || course.course_runs[0].end === null) ? null : (isNaN(Date.parse(course.course_runs[0].end)) ? null : new Date(Date.parse(course.course_runs[0].end)).getFullYear() + "-" + (parseInt(new Date(Date.parse(course.course_runs[0].end)).getMonth()) + 1) + "-" + new Date(Date.parse(course.course_runs[0].end)).getDate()),
          "SelfPaced": (typeof course.course_runs[0].pacing_type === 'undefined' || course.course_runs[0].pacing_type === null || course.course_runs[0].pacing_type !== 'self_paced') ? false : true,
          "Difficulty": typeof course.course_runs[0].level_type === 'undefined' ? null : course.course_runs[0].level_type
        }));
      if (catalogData.next != null) {
        // console.log("Length: ", courses.length, " :: Next: ", catalogData.next);
        return helpers.doRecursiveRequest(catalogData.next, accessToken, courses);
        // return helpers.doRecursiveRequest("https://prod-edx-discovery.edx.org/api/v1/catalogs/420/courses/?limit=20&offset=1300", accessToken, courses);
      } else {
        return courses;
      }
    }).then(courses => {
      console.log("Courses Length: ", courses.length);
      return helpers.doRecursiveInsert(courses, 0, 0);
    }).then(finalResponse => {
      if (finalResponse.status === 222) {
        resolve({ status: 200, body: JSON.stringify({ statusText: 'Success', message: finalResponse.body.recordsProcessed + ' records processed | ' + finalResponse.body.recordsInserted + ' records inserted' }) });
      } else {
        throw Error("Insert Failed Due to an Unknown Reason");
      }
    }).catch(error => {
      reject(error);
    });
  });
};